src/shogun/optimization/FirstOrderMinimizer.h

```diff
@@ -0,0 +1,161 @@
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ * The views and conclusions contained in the software and documentation are those
+ * of the authors and should not be interpreted as representing official policies,
+ * either expressed or implied, of the Shogun Development Team.
+ *
+ */
+
+#ifndef FIRSTORDERMINIMIZER_H
+#define FIRSTORDERMINIMIZER_H
+#include <shogun/optimization/FirstOrderCostFunction.h>
+#include <shogun/optimization/MinimizerContext.h>
+#include <shogun/optimization/Penalty.h>
+namespace shogun
+{
+
+/** @brief The first order minimizer base class.
+ *
+ * This class gives the interface of a minimizer
+ *
+ */
+class CFirstOrderMinimizer
+{
+public: 
+	/** Default constructor */
+	CFirstOrderMinimizer()
+	{
+		init();
+	}
+	/** Constructor
+	 * @param fun cost function
+	 */
+	CFirstOrderMinimizer(CFirstOrderCostFunction *fun)
+	{
+		init();
+		set_cost_function(fun);
+	}
+
+	/** Destructor */
+	virtual ~CFirstOrderMinimizer()
+	{}
+
+	/** Do minimization and get the optimal value 
+	 * 
+	 * @return optimal value
+	 */
+	virtual float64_t minimize()=0;
+
+	/** Does minimizer support batch update
+	 * 
+	 * @return whether minimizer supports batch update
+	 */
+	virtual bool supports_batch_update() const=0;
+
+	/** Set cost function used in the minimizer
+	 *
+	 * @param fun the cost function
+	 */
+	virtual void set_cost_function(CFirstOrderCostFunction *fun)
+	{
+		m_fun=fun;
+	}
+
+	virtual CMinimizerContext* save_to_context()=0;
+
+	virtual void load_from_context(CMinimizerContext* context)=0;
+
+
+	/** set the weight of penalty
+	 * for example, for L2 penalty
+	 * the penalized cost function is =unpenalized cost function+penalty_weight*L2_penalty
+	 *
+	 * @param penalty_weight the weight of penalty
+	 */
+	virtual void set_penalty_weight(float64_t penalty_weight)
+	{
+		//REQUIRE(penalty_weight>0,"penalty_weight must be positive\n");
+		m_penalty_weight=penalty_weight;
+	}
+
+	/** set the type of penalty
+	 * for example, L2 penalty
+	 *
+	 * @param penalty_type the type of penalty
+	 */
+	virtual void set_penalty_type(CPenalty* penalty_type)
+	{
+		//REQUIRE(penalty_type,"the type of penalty_type must not be NULL\n");
+		if(m_penalty_type!=penalty_type)
+		{
+			m_penalty_type=penalty_type;
+		}
+	}
+protected:
+	virtual float64_t get_penalty(SGVector<float64_t> var)
+	{
+		float64_t penalty=0.0;
+		if(m_penalty_type)
+		{
+			for(auto idx=0; idx<var.vlen; idx++)
```

Comment:
@karlnapf 
travis complains about the `auto` keyword.
ref
https://travis-ci.org/shogun-toolbox/shogun/jobs/73320014

In fact, we can use the `linalg` here.